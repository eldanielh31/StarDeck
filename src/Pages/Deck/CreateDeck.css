.containerDeck {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.containeritemMenu {
    width: 100%;
    display: flex;
    flex-direction: row;
}

.containermenuDeck {
    padding: 0.5em;
    width: 95%;
    display: flex;
    flex-direction: row;
    border-bottom: 1px solid black;
}

.rigthItemMenu {
    float: right;
    width: 25%;
    justify-content: left;
    display: flex;
    align-items: center;
}

.leftItemMenu {
    float: left;
    width: 25%;
    justify-content: right;
    display: flex;
    align-items: center;
}

.centerItemMenu {
    float: center;
    width: 50%;
    justify-content: center;
    align-items: center;
    display: flex;
}

.inputNameDeck {
    width: 100%;
    border: none;
    border-bottom: 2px inset black;
    border-radius: 10 px;
    outline: none;
    font-size: 20px;
    text-align: center;
}

.containerCards {
    height: 90vh;
    overflow: scroll;
    display: flex;
    flex-wrap: wrap;
}

/* CSS */
.button-57 {
    margin: 1em;
    position: relative;
    overflow: hidden;
    border: 1px solid #18181a;
    border-radius: 5px;
    color: #18181a;
    display: inline-block;
    font-size: 15px;
    line-height: 15px;
    padding: 10px 30px 10px;
    text-decoration: none;
    cursor: pointer;
    background: #fff;
    user-select: none;
    -webkit-user-select: none;
    touch-action: manipulation;
}

.button-57 span:first-child {
    position: relative;
    transition: color 600ms cubic-bezier(0.48, 0, 0.12, 1);
    z-index: 10;
}

.button-57 span:last-child {
    color: white;
    display: block;
    position: absolute;
    bottom: 0;
    transition: all 500ms cubic-bezier(0.48, 0, 0.12, 1);
    z-index: 100;
    opacity: 0;
    top: 50%;
    left: 50%;
    transform: translateY(225%) translateX(-50%);
    height: 14px;
    line-height: 13px;
}

.button-57:after {
    content: "";
    position: absolute;
    bottom: -50%;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: black;
    transform-origin: bottom center;
    transition: transform 600ms cubic-bezier(0.48, 0, 0.12, 1);
    transform: skewY(9.3deg) scaleY(0);
    z-index: 50;
}

.button-57:hover:after {
    transform-origin: bottom center;
    transform: skewY(9.3deg) scaleY(2);
}

.button-57:hover span:last-child {
    transform: translateX(-50%) translateY(-100%);
    opacity: 1;
    transition: all 900ms cubic-bezier(0.48, 0, 0.12, 1);
}