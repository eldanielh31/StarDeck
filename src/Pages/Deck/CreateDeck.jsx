import React, { useEffect, useState } from 'react'

import './CreateDeck.css'
import { createDeck, getCartas, getDeck } from '../../requestApi';
import { v4 as uuidv4 } from "uuid";
import { theme } from '../../theme';
import { Alert } from '@mui/material';


function CreateDeck() {

    const [cards, setCards] = useState([]);
    const [cartasSeleccionadas, setCartasSeleccionadas] = useState([]);
    const [name, setName] = useState('');
    const [code, setCode] = useState('');
    const [decks, setDecks] = useState([]);

    const [alert, setAlert] = useState(null);
    const [textAlert, settextAlert] = useState('');
    const [typeAlert, settypeAlert] = useState('');

    /* istanbul ignore next */
    const handleSubmit = (e) => {
        e.preventDefault();

        if (!(cartasSeleccionadas.length === 18)) {
            setAlert(true);
            settypeAlert('error');
            settextAlert('Debe ingresar 18 cartas');
        } else if (existsName(name)) {
            setAlert(true);
            settypeAlert('error');
            settextAlert('El nombre ya existe');
        }
        else {
            sendDeck();
        }

    };

    /* istanbul ignore next */
    const existsName = (name) => {
        return decks.some(deck => deck.name === name);
    };

    /* istanbul ignore next */
    const sendDeck = () => {

        const codigo = 'C-' + uuidv4().split('').reverse().join('').slice(0, 12);

        setCode(codigo);
        let baraja = {
            name: name,
            cards: cartasSeleccionadas,
            id: code,
            id_User: 1
        }

        createDeck(baraja).then(
            () => {
                console.log("Enviado correcto.");
                setDecks([...decks,  baraja]);
                setAlert(true);
                settypeAlert('success');
                textAlert('Guardado Correctamente');
            }
        );

    }

    /* istanbul ignore next */
    const manejarSeleccionDeCarta = (id) => {
        if (cartasSeleccionadas.includes(id)) {
            setCartasSeleccionadas(cartasSeleccionadas.filter((c) => c !== id));
        } else {
            setCartasSeleccionadas([...cartasSeleccionadas, id]);
        }
    };

    /* istanbul ignore next */
    useEffect(() => {
        
        getCartas().then(
            (findCards) => {
                setCards(findCards);
            }
        );

        getDeck().then(
            (findDeck) => {
                setDecks(findDeck)
            }
        );

        function confirmLeave(e) {
            e.preventDefault();
            e.returnValue = '¿Seguro que deseas abandonar esta página? Los cambios que hayas realizado se perderán.';
        }

        window.addEventListener('beforeunload', confirmLeave);

        return () => {
            window.removeEventListener('beforeunload', confirmLeave);
        };

    }, [setCards, setDecks]);

    /* istanbul ignore next */
    return (
        <>
            <div className='containerDeck'>
                <div className='containermenuDeck'>

                    <form className='containeritemMenu' onSubmit={handleSubmit}>
                        <div className='rigthItemMenu'>
                            <h3>Cartas({cards?.length})</h3>
                        </div>

                        <div className='centerItemMenu'>
                            <input minLength={5} maxLength={30} className='inputNameDeck' onChange={(e) => setName(e.target.value)} type="text" name="name" placeholder='Inserte el nombre de la baraja' id="" required />
                            {alert && <Alert severity={typeAlert}>{textAlert}</Alert>}
                        </div>

                        <div className='leftItemMenu'>
                            <h3>Seleccionadas({cartasSeleccionadas.length})</h3>
                            <button className="button-57" type='submit'><span className="text">Crear</span><span>Ok!</span></button>
                        </div>
                    </form>

                </div>
                <div className='containerCards'>
                    {cards?.map((card) => (
                        <div key={card.id} value={card.id} onClick={() => manejarSeleccionDeCarta(card.id)} style={{
                            textAlign: "center", 
                            height: "20em", margin: 
                            "10px", padding: "10px", 
                            border: `5px solid ${cartasSeleccionadas.includes(card.id) ? "red" : "black"}`, 
                            borderRadius: "5px",
                            background: `linear-gradient(to top, ${theme.colors.primary}, ${theme.colors.secondary}`,
                        }}>
                            <img src={card.imagen} width={200} height={250} alt="Imagen" />
                            <h3 style={{color: 'white' }}>{card.nombre}</h3>
                        </div>
                    ))}
                </div>

            </div>

        </>

    )
}

export default CreateDeck