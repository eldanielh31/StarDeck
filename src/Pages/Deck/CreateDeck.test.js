import React from 'react';
import { render } from '@testing-library/react';
import CreateDeck from './CreateDeck.jsx';


describe('Pantalla Deck', () => {

    // Prueba de renderizado
    it('Renderiza correctamente', () => {
      const { container } = render(<CreateDeck />);
      expect(container.firstChild).toMatchSnapshot();
    });
  
    // Prueba de propiedades
    it('Muestra el texto adecuado', () => {
      const { getByText, getAllByText } = render(<CreateDeck/>);
      expect(getAllByText('Crear')[0]).toBeInTheDocument();
      
    });
  
  });