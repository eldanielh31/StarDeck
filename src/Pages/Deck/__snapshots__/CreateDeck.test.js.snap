// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pantalla Deck Renderiza correctamente 1`] = `
<div
  class="containerDeck"
>
  <div
    class="containermenuDeck"
  >
    <form
      class="containeritemMenu"
    >
      <div
        class="rigthItemMenu"
      >
        <h3>
          Cartas(
          0
          )
        </h3>
      </div>
      <div
        class="centerItemMenu"
      >
        <input
          class="inputNameDeck"
          id=""
          maxlength="30"
          minlength="5"
          name="name"
          placeholder="Inserte el nombre de la baraja"
          required=""
          type="text"
        />
      </div>
      <div
        class="leftItemMenu"
      >
        <h3>
          Seleccionadas(
          0
          )
        </h3>
        <button
          class="button-57"
          type="submit"
        >
          <span
            class="text"
          >
            Crear
          </span>
          <span>
            Ok!
          </span>
        </button>
      </div>
    </form>
  </div>
  <div
    class="containerCards"
  />
</div>
`;
